<a href="https://colab.research.google.com/github/sergiocostaifes/PPCOMP_DM/blob/main/notebooks/04_window_5min_episodes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NB04 — Detecção de Episódios Críticos

## 1. Contexto

A partir da série temporal contínua construída no NB03, este notebook realiza a identificação de janelas críticas e sua consolidação em episódios críticos.

O NB03 produz o arquivo `window_5min_series.parquet`, contendo a série contínua em janelas de 5 minutos, com métricas agregadas de eventos, falhas, máquinas, coleções, prioridade e requisições de recursos. O NB04 utiliza essa série para calcular a taxa de falha por janela e identificar períodos de criticidade operacional.

A análise deixa de considerar eventos isolados e passa a focar em padrões agregados ao longo do tempo, permitindo identificar períodos de degradação operacional com base em uma métrica relativa de falha por janela.

Como a execução atual do NB03 preservou o início da série e confirmou a presença de `bucket_id = 0`, esta etapa também documenta explicitamente o comportamento do bucket inicial e sua influência sobre os parâmetros estatísticos globais.

Além disso, como o critério oficial de criticidade utiliza um limiar calculado sobre toda a série temporal, este notebook inclui uma análise de sensibilidade com limiar estimado apenas no trecho inicial da série, correspondente ao período de treinamento temporal.

## 2. Objetivo

O objetivo principal deste notebook é detectar janelas críticas e consolidá-las em episódios críticos.

Especificamente, este notebook busca:

1. Ler a série contínua `window_5min_series.parquet`;
2. Validar a estrutura da série temporal contínua;
3. Calcular a taxa de falha por janela (`fail_rate`);
4. Tratar janelas sem eventos definindo `fail_rate = 0`;
5. Calcular o limiar oficial global de criticidade com base em `μ + 2σ`;
6. Classificar janelas críticas no cenário oficial;
7. Consolidar janelas críticas contíguas em episódios críticos;
8. Calcular um limiar alternativo usando apenas os primeiros 80% da série;
9. Comparar o cenário oficial global com o cenário de sensibilidade;
10. Diagnosticar o comportamento do `bucket_id = 0`;
11. Avaliar o efeito da remoção diagnóstica do `bucket_id = 0`;
12. Persistir a série enriquecida com `fail_rate` e `is_critical`;
13. Persistir a tabela de episódios críticos;
14. Gerar summaries JSON oficiais e de sensibilidade.

## 3. Papel no Pipeline

O NB04 desempenha papel central na estruturação semântica do pipeline.

Ele transforma a série temporal agregada em uma representação segmentada composta por:

1. janelas não críticas;
2. janelas críticas;
3. episódios críticos delimitados por contiguidade temporal.

Essa estrutura fornece a referência a partir da qual, nos notebooks posteriores, serão definidos estados operacionais, construídos atributos temporais e formulado o problema preditivo.

| Notebook | Relação com o NB04 |
|---|---|
| NB03 | Fornece `window_5min_series.parquet` |
| NB04 | Calcula `fail_rate`, classifica janelas críticas e detecta episódios |
| NB05 | Usa a série enriquecida para engenharia de atributos |
| NB06 | Usa os episódios e a série enriquecida para estados e rotulagem |
| NB08/NB09 | Usam a formulação supervisionada derivada das etapas anteriores |

## 4. Métrica de Detecção

A métrica adotada para caracterizar criticidade em cada janela é:

\[
fail\_rate_t = \frac{n\_failed_t}{n\_events_t}
\]

Para janelas com `n_events = 0`, define-se:

\[
fail\_rate_t = 0
\]

Essa decisão preserva a continuidade da série temporal e evita valores indefinidos no fluxo principal de análise.

A métrica baseada em taxa reduz o viés associado ao volume absoluto de eventos, pois avalia a proporção relativa de falhas em cada janela.

## 5. Critério Oficial de Criticidade

A identificação oficial de janelas críticas é realizada por meio de um limiar estatístico global baseado na distribuição da taxa de falha:

\[
threshold = \mu + 2\sigma
\]

em que:

- \(\mu\) representa a média global de `fail_rate`;
- \(\sigma\) representa o desvio-padrão global de `fail_rate`.

Uma janela é classificada como crítica quando:

\[
fail\_rate_t \geq threshold
\]

Esse critério global permanece como regra oficial retrospectiva do pipeline.

## 6. Análise de Sensibilidade do Limiar

Além da regra oficial, este notebook executa uma análise de sensibilidade com limiar calculado apenas sobre os primeiros 80% da série temporal.

Essa análise simula uma condição mais próxima de uso operacional, na qual estatísticas futuras ainda não estariam disponíveis para o cálculo de \(\mu\) e \(\sigma\).

São comparados dois cenários:

| Cenário | Descrição |
|---|---|
| `global_full_series` | Limiar calculado sobre toda a série temporal |
| `train_only_threshold` | Limiar calculado apenas sobre os primeiros 80% da série |

A comparação considera:

1. valor de \(\mu\);
2. valor de \(\sigma\);
3. valor do limiar;
4. número de janelas críticas;
5. número de episódios detectados;
6. diferença em relação ao cenário oficial global.

A análise de sensibilidade não substitui automaticamente a segmentação oficial. Ela serve para avaliar a robustez metodológica da escolha do limiar.

## 7. Construção de Episódios

As janelas classificadas como críticas são agrupadas em intervalos contíguos, formando episódios críticos.

Cada episódio representa um período contínuo de comportamento anômalo ou degradação operacional, sendo caracterizado por:

1. bucket inicial;
2. bucket final;
3. duração em janelas;
4. duração em minutos;
5. maior taxa de falha;
6. taxa média de falha;
7. maior número absoluto de falhas;
8. média de falhas;
9. soma de falhas;
10. maior volume de eventos;
11. média de eventos.

Essa consolidação permite tratar a criticidade como fenômeno temporal estruturado, e não apenas como ocorrência pontual.

## 8. Artefatos Utilizados e Gerados

Este notebook utiliza como entrada:

| Artefato | Origem | Finalidade |
|---|---|---|
| `window_5min_series.parquet` | NB03 | Série temporal contínua em janelas de 5 minutos |

O notebook gera os artefatos oficiais:

| Artefato | Finalidade |
|---|---|
| `window_5min_series_scored.parquet` | Série enriquecida com `fail_rate` e `is_critical` no cenário oficial |
| `episodes_detected.parquet` | Episódios críticos detectados pelo limiar global |
| `04_detect_episodes_summary.json` | Summary oficial da detecção de episódios |

Também são gerados artefatos de sensibilidade:

| Artefato | Finalidade |
|---|---|
| `window_5min_series_scored_train_threshold.parquet` | Série classificada com limiar estimado no trecho de treino |
| `episodes_detected_train_threshold.parquet` | Episódios detectados com limiar de treino |
| `04_threshold_sensitivity_summary.json` | Summary comparativo da sensibilidade do limiar |

## 9. Observações Metodológicas

O limiar global é adequado para uma análise retrospectiva e reprodutível do dataset, pois define um critério único de criticidade para toda a série analisada.

Entretanto, em uso operacional estritamente online, estatísticas futuras não estariam disponíveis no momento da decisão. Por isso, a análise de sensibilidade com limiar calculado apenas sobre o trecho inicial da série é mantida como evidência metodológica.

A análise sem o `bucket_id = 0` é diagnóstica. Ela não substitui a regra oficial do notebook, mas ajuda a avaliar se o bucket inicial distorce substancialmente a segmentação episódica.

O arquivo `window_5min_series_scored.parquet` e a tabela `episodes_detected.parquet` devem ser considerados os artefatos oficiais desta etapa para os notebooks posteriores.

In [ ]:
# ============================================================
# NB04 — Detecção de Episódios Críticos
# Pipeline PPCOMP_DM
#
# Escopo:
# - Ler a série contínua produzida pelo NB03
# - Calcular fail_rate por janela
# - Classificar janelas críticas pelo critério global mu + 2*sigma
# - Consolidar janelas críticas contíguas em episódios críticos
# - Executar análise de sensibilidade com limiar calculado no trecho de treino
# - Diagnosticar o comportamento do bucket_id = 0
# - Persistir artefatos oficiais e de sensibilidade
# - Persistir summaries JSON oficiais e comparativos
#
# Importante:
# - O cenário global é a segmentação oficial retrospectiva do pipeline
# - O cenário train-only é uma análise de sensibilidade
# - A remoção de bucket_id = 0 é apenas diagnóstico complementar
# - Este notebook não realiza engenharia de atributos nem modelagem
# ============================================================

from pathlib import Path
import os
import sys
import subprocess
import importlib
import random
import json

import numpy as np
import pandas as pd
from IPython.display import display

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

# ─────────────────────────────────────────────────────────────
# BLOCO 0 — Bootstrap
# ─────────────────────────────────────────────────────────────

if not Path("/content/drive/MyDrive").exists():
    from google.colab import drive
    drive.mount("/content/drive")
else:
    print("[Bootstrap] Google Drive já montado.")

REPO_DIR = Path("/content/drive/MyDrive/Mestrado/PPCOMP_DM")
GITHUB_REPO = "https://github.com/sergiocostaifes/PPCOMP_DM.git"

# Garante o repositório local no Google Drive, clonando apenas se necessário.
if not REPO_DIR.exists():
    REPO_DIR.parent.mkdir(parents=True, exist_ok=True)
    print(f"[Bootstrap] Clonando repositório em: {REPO_DIR}")
    subprocess.run(["git", "clone", GITHUB_REPO, str(REPO_DIR)], check=True)
else:
    try:
        print("[Bootstrap] Atualizando repositório (git pull)...")
        subprocess.run(["git", "-C", str(REPO_DIR), "pull"], check=True)
    except Exception as e:
        print("[Bootstrap] Aviso: não foi possível atualizar via git pull:", e)

os.chdir(str(REPO_DIR))
print("[Bootstrap] CWD =", os.getcwd())

repo_str = str(REPO_DIR)
if repo_str not in sys.path:
    sys.path.insert(0, repo_str)

importlib.invalidate_caches()

from src.paths import FEATURES_PATH, REPORTS_PATH, ensure_dirs

ensure_dirs()

print("FEATURES_PATH =", FEATURES_PATH)
print("REPORTS_PATH =", REPORTS_PATH)

def log(msg: str) -> None:
    """Imprime mensagens padronizadas desta etapa do pipeline."""
    print(f"[NB04_detect_episodes] {msg}")

# ─────────────────────────────────────────────────────────────
# BLOCO 1 — Parâmetros
# ─────────────────────────────────────────────────────────────

WINDOW_MINUTES = 5
TRAIN_FRACTION_FOR_THRESHOLD = 0.80

SERIES_FILE = FEATURES_PATH / "window_5min_series.parquet"

OUT_SERIES_SCORED_GLOBAL = FEATURES_PATH / "window_5min_series_scored.parquet"
OUT_EPISODES_GLOBAL = FEATURES_PATH / "episodes_detected.parquet"

OUT_SERIES_SCORED_TRAIN_ONLY = (
    FEATURES_PATH / "window_5min_series_scored_train_threshold.parquet"
)
OUT_EPISODES_TRAIN_ONLY = (
    FEATURES_PATH / "episodes_detected_train_threshold.parquet"
)

SUMMARY_FILE = REPORTS_PATH / "04_detect_episodes_summary.json"
SENSITIVITY_SUMMARY_FILE = REPORTS_PATH / "04_threshold_sensitivity_summary.json"

# ─────────────────────────────────────────────────────────────
# BLOCO 2 — Leitura da série contínua
# ─────────────────────────────────────────────────────────────

assert SERIES_FILE.exists(), f"Arquivo não encontrado: {SERIES_FILE}"

df_base = pd.read_parquet(SERIES_FILE)

required_cols = {"bucket_id", "bucket_start_us", "n_failed", "n_events"}
missing = required_cols - set(df_base.columns)

assert not missing, f"Colunas ausentes na série: {missing}"

df_base = df_base.sort_values("bucket_id").reset_index(drop=True)

assert df_base["bucket_id"].is_monotonic_increasing, (
    "bucket_id não está ordenado crescentemente."
)

assert df_base["bucket_id"].is_unique, (
    "Foram encontrados bucket_id duplicados."
)

bucket_diff = df_base["bucket_id"].diff().dropna()
n_internal_gaps_gt1 = int((bucket_diff > 1).sum())

assert n_internal_gaps_gt1 == 0, (
    "A série contínua deveria estar sem lacunas internas após o NB03."
)

series_rows = int(len(df_base))
bucket_min = int(df_base["bucket_id"].min())
bucket_max = int(df_base["bucket_id"].max())

log(f"Série contínua carregada: rows={series_rows}, bucket range={bucket_min}..{bucket_max}")

# ─────────────────────────────────────────────────────────────
# BLOCO 3 — Cálculo da taxa de falha
# ─────────────────────────────────────────────────────────────

df_base = df_base.copy()

zero_event_windows = int((df_base["n_events"] == 0).sum())
nonzero_event_windows = int((df_base["n_events"] > 0).sum())

den = df_base["n_events"].replace(0, np.nan)

df_base["fail_rate"] = (
    df_base["n_failed"] / den
).fillna(0.0).astype("float32")

log(f"Janelas com n_events == 0: {zero_event_windows}")
log(f"Janelas com n_events > 0: {nonzero_event_windows}")

# ─────────────────────────────────────────────────────────────
# BLOCO 4 — Funções auxiliares
# ─────────────────────────────────────────────────────────────

def compute_threshold(values: pd.Series) -> dict:
    """Calcula média, desvio-padrão populacional e limiar mu + 2*sigma."""
    mu = float(values.mean())
    sigma = float(values.std(ddof=0))
    threshold = float(mu + 2 * sigma)

    return {
        "mu": mu,
        "sigma": sigma,
        "threshold": threshold,
    }


def detect_episodes_from_flag(
    df: pd.DataFrame,
    critical_col: str,
    threshold_info: dict,
    scenario_name: str,
) -> pd.DataFrame:
    """
    Consolida janelas críticas contíguas em episódios.

    Cada episódio é uma sequência máxima de linhas consecutivas com critical_col == 1.
    """
    episodes = []
    in_episode = False
    start_idx = None

    flags = df[critical_col].to_numpy()

    for i, flag in enumerate(flags):
        if flag == 1 and not in_episode:
            in_episode = True
            start_idx = i

        elif flag == 0 and in_episode:
            episodes.append((start_idx, i - 1))
            in_episode = False

    if in_episode:
        episodes.append((start_idx, len(df) - 1))

    rows = []

    for ep_id, (s, e) in enumerate(episodes, start=1):
        seg = df.iloc[s:e + 1]

        rows.append({
            "scenario": scenario_name,
            "episode_id": ep_id,
            "start_bucket": int(seg["bucket_id"].iloc[0]),
            "end_bucket": int(seg["bucket_id"].iloc[-1]),
            "start_bucket_start_us": int(seg["bucket_start_us"].iloc[0]),
            "end_bucket_start_us": int(seg["bucket_start_us"].iloc[-1]),
            "duration_windows": int(len(seg)),
            "duration_minutes": int(len(seg) * WINDOW_MINUTES),
            "max_fail_rate": float(seg["fail_rate"].max()),
            "mean_fail_rate": float(seg["fail_rate"].mean()),
            "max_failed": int(seg["n_failed"].max()),
            "mean_failed": float(seg["n_failed"].mean()),
            "sum_failed": int(seg["n_failed"].sum()),
            "max_events": int(seg["n_events"].max()),
            "mean_events": float(seg["n_events"].mean()),
            "threshold": float(threshold_info["threshold"]),
            "mu": float(threshold_info["mu"]),
            "sigma": float(threshold_info["sigma"]),
        })

    return pd.DataFrame(rows)


def summarize_episode_result(
    df_scored: pd.DataFrame,
    episodes_df: pd.DataFrame,
    critical_col: str,
    threshold_info: dict,
    scenario_name: str,
) -> dict:
    """Resume a classificação crítica e a distribuição de duração dos episódios."""
    critical_windows = int(df_scored[critical_col].sum())
    episodes_detected = int(len(episodes_df))

    if episodes_detected > 0:
        one_window_episodes = int((episodes_df["duration_windows"] == 1).sum())
        duration_windows_mean = float(episodes_df["duration_windows"].mean())
        duration_windows_median = float(episodes_df["duration_windows"].median())
        duration_windows_max = int(episodes_df["duration_windows"].max())
        duration_minutes_mean = float(episodes_df["duration_minutes"].mean())
        duration_minutes_median = float(episodes_df["duration_minutes"].median())
        duration_minutes_max = int(episodes_df["duration_minutes"].max())
    else:
        one_window_episodes = 0
        duration_windows_mean = 0.0
        duration_windows_median = 0.0
        duration_windows_max = 0
        duration_minutes_mean = 0.0
        duration_minutes_median = 0.0
        duration_minutes_max = 0

    return {
        "scenario": scenario_name,
        "mu": float(threshold_info["mu"]),
        "sigma": float(threshold_info["sigma"]),
        "threshold": float(threshold_info["threshold"]),
        "critical_windows": critical_windows,
        "episodes_detected": episodes_detected,
        "one_window_episodes": one_window_episodes,
        "duration_windows_mean": duration_windows_mean,
        "duration_windows_median": duration_windows_median,
        "duration_windows_max": duration_windows_max,
        "duration_minutes_mean": duration_minutes_mean,
        "duration_minutes_median": duration_minutes_median,
        "duration_minutes_max": duration_minutes_max,
    }

# ─────────────────────────────────────────────────────────────
# BLOCO 5 — Cenário oficial: limiar global
# ─────────────────────────────────────────────────────────────

global_threshold_info = compute_threshold(df_base["fail_rate"])

df_global = df_base.copy()

df_global["is_critical"] = (
    df_global["fail_rate"] >= global_threshold_info["threshold"]
).astype("int8")

episodes_global = detect_episodes_from_flag(
    df=df_global,
    critical_col="is_critical",
    threshold_info=global_threshold_info,
    scenario_name="global_full_series",
)

summary_global = summarize_episode_result(
    df_scored=df_global,
    episodes_df=episodes_global,
    critical_col="is_critical",
    threshold_info=global_threshold_info,
    scenario_name="global_full_series",
)

log("[Cenário oficial — global]")
log(f"mu={summary_global['mu']:.8f}")
log(f"sigma={summary_global['sigma']:.8f}")
log(f"threshold={summary_global['threshold']:.8f}")
log(f"critical_windows={summary_global['critical_windows']}")
log(f"episodes_detected={summary_global['episodes_detected']}")

# ─────────────────────────────────────────────────────────────
# BLOCO 6 — Cenário de sensibilidade: limiar calculado no treino
# ─────────────────────────────────────────────────────────────

train_cutoff_idx = int(np.floor(series_rows * TRAIN_FRACTION_FOR_THRESHOLD))
train_cutoff_idx = max(1, min(train_cutoff_idx, series_rows))

df_train_portion = df_base.iloc[:train_cutoff_idx].copy()

train_threshold_info = compute_threshold(df_train_portion["fail_rate"])

df_train_threshold = df_base.copy()

df_train_threshold["is_critical"] = (
    df_train_threshold["fail_rate"] >= train_threshold_info["threshold"]
).astype("int8")

episodes_train_threshold = detect_episodes_from_flag(
    df=df_train_threshold,
    critical_col="is_critical",
    threshold_info=train_threshold_info,
    scenario_name="train_only_threshold",
)

summary_train_threshold = summarize_episode_result(
    df_scored=df_train_threshold,
    episodes_df=episodes_train_threshold,
    critical_col="is_critical",
    threshold_info=train_threshold_info,
    scenario_name="train_only_threshold",
)

log("[Cenário sensibilidade — limiar calculado no treino]")
log(f"train_fraction={TRAIN_FRACTION_FOR_THRESHOLD}")
log(f"train_cutoff_idx={train_cutoff_idx}")
log(f"mu={summary_train_threshold['mu']:.8f}")
log(f"sigma={summary_train_threshold['sigma']:.8f}")
log(f"threshold={summary_train_threshold['threshold']:.8f}")
log(f"critical_windows={summary_train_threshold['critical_windows']}")
log(f"episodes_detected={summary_train_threshold['episodes_detected']}")

# ─────────────────────────────────────────────────────────────
# BLOCO 7 — Comparação entre cenários
# ─────────────────────────────────────────────────────────────

comparison_df = pd.DataFrame([
    summary_global,
    summary_train_threshold,
])

comparison_df["delta_critical_windows_vs_global"] = (
    comparison_df["critical_windows"] - summary_global["critical_windows"]
)

comparison_df["delta_episodes_vs_global"] = (
    comparison_df["episodes_detected"] - summary_global["episodes_detected"]
)

comparison_df["delta_threshold_vs_global"] = (
    comparison_df["threshold"] - summary_global["threshold"]
)

print("\n=== Comparação entre limiares ===")
display(comparison_df)

# ─────────────────────────────────────────────────────────────
# BLOCO 8 — Diagnóstico do bucket 0
# ─────────────────────────────────────────────────────────────

bucket0_present = bool((df_global["bucket_id"] == 0).any())
bucket0_summary = None

if bucket0_present:
    row0_global = df_global.loc[df_global["bucket_id"] == 0].iloc[0]
    row0_train_threshold = df_train_threshold.loc[
        df_train_threshold["bucket_id"] == 0
    ].iloc[0]

    bucket0_summary = {
        "bucket_id": int(row0_global["bucket_id"]),
        "bucket_start_us": int(row0_global["bucket_start_us"]),
        "n_events": int(row0_global["n_events"]),
        "n_failed": int(row0_global["n_failed"]),
        "fail_rate": float(row0_global["fail_rate"]),
        "is_critical_global": int(row0_global["is_critical"]),
        "is_critical_train_threshold": int(row0_train_threshold["is_critical"]),
    }

    log(f"Bucket 0 summary: {bucket0_summary}")
else:
    log("Bucket 0 não encontrado na série contínua.")

# ─────────────────────────────────────────────────────────────
# BLOCO 9 — Diagnóstico complementar sem bucket 0
# ─────────────────────────────────────────────────────────────

diag_without_bucket0 = None

if bucket0_present and len(df_base) > 1:
    df_wo_b0 = df_base[df_base["bucket_id"] != 0].copy()

    threshold_wo_b0 = compute_threshold(df_wo_b0["fail_rate"])

    df_wo_b0["is_critical"] = (
        df_wo_b0["fail_rate"] >= threshold_wo_b0["threshold"]
    ).astype("int8")

    episodes_wo_b0 = detect_episodes_from_flag(
        df=df_wo_b0,
        critical_col="is_critical",
        threshold_info=threshold_wo_b0,
        scenario_name="without_bucket0_diagnostic",
    )

    diag_without_bucket0 = summarize_episode_result(
        df_scored=df_wo_b0,
        episodes_df=episodes_wo_b0,
        critical_col="is_critical",
        threshold_info=threshold_wo_b0,
        scenario_name="without_bucket0_diagnostic",
    )

    log("[Diagnóstico complementar — sem bucket 0]")
    log(f"mu={diag_without_bucket0['mu']:.8f}")
    log(f"sigma={diag_without_bucket0['sigma']:.8f}")
    log(f"threshold={diag_without_bucket0['threshold']:.8f}")
    log(f"critical_windows={diag_without_bucket0['critical_windows']}")
    log(f"episodes_detected={diag_without_bucket0['episodes_detected']}")

# ─────────────────────────────────────────────────────────────
# BLOCO 10 — Persistência dos artefatos
# ─────────────────────────────────────────────────────────────

df_global.to_parquet(
    OUT_SERIES_SCORED_GLOBAL,
    compression="snappy",
    index=False,
)

episodes_global.to_parquet(
    OUT_EPISODES_GLOBAL,
    compression="snappy",
    index=False,
)

df_train_threshold.to_parquet(
    OUT_SERIES_SCORED_TRAIN_ONLY,
    compression="snappy",
    index=False,
)

episodes_train_threshold.to_parquet(
    OUT_EPISODES_TRAIN_ONLY,
    compression="snappy",
    index=False,
)

log(f"Série enriquecida oficial salva: {OUT_SERIES_SCORED_GLOBAL}")
log(f"Episódios oficiais salvos: {OUT_EPISODES_GLOBAL}")
log(f"Série com limiar treino salva: {OUT_SERIES_SCORED_TRAIN_ONLY}")
log(f"Episódios com limiar treino salvos: {OUT_EPISODES_TRAIN_ONLY}")

# ─────────────────────────────────────────────────────────────
# BLOCO 11 — Summaries JSON
# ─────────────────────────────────────────────────────────────

summary_official = {
    "notebook": "NB04",
    "purpose": "detect_critical_episodes",
    "metric": "fail_rate",
    "official_threshold_mode": "global_full_series",
    "train_fraction_for_sensitivity": float(TRAIN_FRACTION_FOR_THRESHOLD),
    "input_series_file": str(SERIES_FILE),
    "series_rows": int(series_rows),
    "bucket_id_min": int(bucket_min),
    "bucket_id_max": int(bucket_max),
    "zero_event_windows": int(zero_event_windows),
    "nonzero_event_windows": int(nonzero_event_windows),
    "bucket0_present": bool(bucket0_present),
    "bucket0_summary": bucket0_summary,
    "official_result": summary_global,
    "train_threshold_sensitivity_result": summary_train_threshold,
    "without_bucket0_diagnostic": diag_without_bucket0,
    "output_series_scored_file": str(OUT_SERIES_SCORED_GLOBAL),
    "output_episodes_file": str(OUT_EPISODES_GLOBAL),
    "output_series_scored_train_threshold_file": str(OUT_SERIES_SCORED_TRAIN_ONLY),
    "output_episodes_train_threshold_file": str(OUT_EPISODES_TRAIN_ONLY),
}

threshold_sensitivity_summary = {
    "notebook": "NB04",
    "purpose": "threshold_sensitivity_analysis",
    "metric": "fail_rate",
    "threshold_rule": "mu_plus_2sigma",
    "train_fraction_for_threshold": float(TRAIN_FRACTION_FOR_THRESHOLD),
    "comparison": comparison_df.to_dict("records"),
    "interpretation_note": (
        "The global threshold is the official retrospective criterion. "
        "The train-only threshold is a sensitivity analysis intended to assess "
        "whether the critical episode segmentation is strongly affected by using "
        "future observations to estimate mu and sigma."
    ),
}

SUMMARY_FILE.write_text(
    json.dumps(summary_official, indent=2, ensure_ascii=False),
    encoding="utf-8",
)

SENSITIVITY_SUMMARY_FILE.write_text(
    json.dumps(threshold_sensitivity_summary, indent=2, ensure_ascii=False),
    encoding="utf-8",
)

log(f"Resumo oficial salvo: {SUMMARY_FILE}")
log(f"Resumo de sensibilidade salvo: {SENSITIVITY_SUMMARY_FILE}")

# ─────────────────────────────────────────────────────────────
# BLOCO 12 — Resumo final na tela
# ─────────────────────────────────────────────────────────────

print("\n=== RESUMO FINAL — NB04 ===")
print(f"Série analisada: {series_rows} janelas")
print(f"Bucket range: {bucket_min}..{bucket_max}")
print(f"Janelas sem eventos: {zero_event_windows}")
print()

print("[Oficial — limiar global]")
print(f"mu: {summary_global['mu']:.8f}")
print(f"sigma: {summary_global['sigma']:.8f}")
print(f"threshold: {summary_global['threshold']:.8f}")
print(f"janelas críticas: {summary_global['critical_windows']}")
print(f"episódios: {summary_global['episodes_detected']}")
print()

print("[Sensibilidade — limiar calculado no treino]")
print(f"train_fraction: {TRAIN_FRACTION_FOR_THRESHOLD}")
print(f"mu: {summary_train_threshold['mu']:.8f}")
print(f"sigma: {summary_train_threshold['sigma']:.8f}")
print(f"threshold: {summary_train_threshold['threshold']:.8f}")
print(f"janelas críticas: {summary_train_threshold['critical_windows']}")
print(f"episódios: {summary_train_threshold['episodes_detected']}")
print()

print("[Comparação]")
display(comparison_df)

print("\n[Head episódios oficiais]")
display(episodes_global.head())

print("\n[Head episódios — limiar treino]")
display(episodes_train_threshold.head())

[Bootstrap] Google Drive já montado.
[Bootstrap] Atualizando repositório (git pull)...
[Bootstrap] CWD = /content/drive/MyDrive/Mestrado/PPCOMP_DM
FEATURES_PATH = /content/drive/MyDrive/Mestrado/02-datasets/03-features
REPORTS_PATH = /content/drive/MyDrive/Mestrado/04-reports
[NB04_detect_episodes] Série contínua carregada: rows=8930, bucket range=0..8929
[NB04_detect_episodes] Janelas com n_events == 0: 2
[NB04_detect_episodes] Janelas com n_events > 0: 8928
[NB04_detect_episodes] [Cenário oficial — global]
[NB04_detect_episodes] mu=0.19657958
[NB04_detect_episodes] sigma=0.14091089
[NB04_detect_episodes] threshold=0.47840136
[NB04_detect_episodes] critical_windows=390
[NB04_detect_episodes] episodes_detected=285
[NB04_detect_episodes] [Cenário sensibilidade — limiar calculado no treino]
[NB04_detect_episodes] train_fraction=0.8
[NB04_detect_episodes] train_cutoff_idx=7144
[NB04_detect_episodes] mu=0.18158248
[NB04_detect_episodes] sigma=0.12637807
[NB04_detect_episodes] threshold=0.4

,scenario,mu,sigma,threshold,critical_windows,episodes_detected,one_window_episodes,duration_windows_mean,duration_windows_median,duration_windows_max,duration_minutes_mean,duration_minutes_median,duration_minutes_max,delta_critical_windows_vs_global,delta_episodes_vs_global,delta_threshold_vs_global
0,global_full_series,0.196580,0.140911,0.478401,390,285,227,1.368421,1.0,7,6.842105,5.0,35,0,0,0.000000
1,train_only_threshold,0.181582,0.126378,0.434339,562,388,300,1.448454,1.0,18,7.242268,5.0,90,172,103,-0.044063


[NB04_detect_episodes] Bucket 0 summary: {'bucket_id': 0, 'bucket_start_us': 0, 'n_events': 56452, 'n_failed': 37082, 'fail_rate': 0.6568766236305237, 'is_critical_global': 1, 'is_critical_train_threshold': 1}
[NB04_detect_episodes] [Diagnóstico complementar — sem bucket 0]
[NB04_detect_episodes] mu=0.19652800
[NB04_detect_episodes] sigma=0.14083412
[NB04_detect_episodes] threshold=0.47819625
[NB04_detect_episodes] critical_windows=394
[NB04_detect_episodes] episodes_detected=285
[NB04_detect_episodes] Série enriquecida oficial salva: /content/drive/MyDrive/Mestrado/02-datasets/03-features/window_5min_series_scored.parquet
[NB04_detect_episodes] Episódios oficiais salvos: /content/drive/MyDrive/Mestrado/02-datasets/03-features/episodes_detected.parquet
[NB04_detect_episodes] Série com limiar treino salva: /content/drive/MyDrive/Mestrado/02-datasets/03-features/window_5min_series_scored_train_threshold.parquet
[NB04_detect_episodes] Episódios com limiar treino salvos: /content/drive/MyD

,scenario,mu,sigma,threshold,critical_windows,episodes_detected,one_window_episodes,duration_windows_mean,duration_windows_median,duration_windows_max,duration_minutes_mean,duration_minutes_median,duration_minutes_max,delta_critical_windows_vs_global,delta_episodes_vs_global,delta_threshold_vs_global
0,global_full_series,0.196580,0.140911,0.478401,390,285,227,1.368421,1.0,7,6.842105,5.0,35,0,0,0.000000
1,train_only_threshold,0.181582,0.126378,0.434339,562,388,300,1.448454,1.0,18,7.242268,5.0,90,172,103,-0.044063



[Head episódios oficiais]


,scenario,episode_id,start_bucket,end_bucket,start_bucket_start_us,end_bucket_start_us,duration_windows,duration_minutes,max_fail_rate,mean_fail_rate,max_failed,mean_failed,sum_failed,max_events,mean_events,threshold,mu,sigma
0,global_full_series,1,0,0,0,0,1,5,0.656877,0.656877,37082,37082.0,37082,56452,56452.0,0.478401,0.19658,0.140911
1,global_full_series,2,666,666,199800000000,199800000000,1,5,0.500000,0.500000,7,7.0,7,14,14.0,0.478401,0.19658,0.140911
2,global_full_series,3,967,967,290100000000,290100000000,1,5,0.756757,0.756757,84,84.0,84,111,111.0,0.478401,0.19658,0.140911
3,global_full_series,4,1220,1220,366000000000,366000000000,1,5,0.625000,0.625000,10,10.0,10,16,16.0,0.478401,0.19658,0.140911
4,global_full_series,5,1391,1391,417300000000,417300000000,1,5,0.750000,0.750000,6,6.0,6,8,8.0,0.478401,0.19658,0.140911



[Head episódios — limiar treino]


,scenario,episode_id,start_bucket,end_bucket,start_bucket_start_us,end_bucket_start_us,duration_windows,duration_minutes,max_fail_rate,mean_fail_rate,max_failed,mean_failed,sum_failed,max_events,mean_events,threshold,mu,sigma
0,train_only_threshold,1,0,0,0,0,1,5,0.656877,0.656877,37082,37082.0,37082,56452,56452.0,0.434339,0.181582,0.126378
1,train_only_threshold,2,454,454,136200000000,136200000000,1,5,0.467532,0.467532,36,36.0,36,77,77.0,0.434339,0.181582,0.126378
2,train_only_threshold,3,666,666,199800000000,199800000000,1,5,0.500000,0.500000,7,7.0,7,14,14.0,0.434339,0.181582,0.126378
3,train_only_threshold,4,780,780,234000000000,234000000000,1,5,0.440476,0.440476,37,37.0,37,84,84.0,0.434339,0.181582,0.126378
4,train_only_threshold,5,895,895,268500000000,268500000000,1,5,0.441860,0.441860,19,19.0,19,43,43.0,0.434339,0.181582,0.126378


# Conclusão da Detecção de Episódios Críticos

A etapa de detecção de episódios críticos foi concluída com sucesso.

O NB04 leu a série contínua `window_5min_series.parquet`, calculou a taxa de falha por janela, aplicou o critério oficial de criticidade baseado em `μ + 2σ` e consolidou janelas críticas contíguas em episódios críticos.

Nesta etapa também foi executada uma análise de sensibilidade com limiar calculado apenas sobre os primeiros 80% da série temporal, além de um diagnóstico complementar sobre o efeito do `bucket_id = 0`.

## 10. Resultados da Execução Oficial

A série contínua analisada apresentou:

| Item | Valor |
|---|---:|
| Total de janelas | 8.930 |
| `bucket_id` mínimo | 0 |
| `bucket_id` máximo | 8.929 |
| Janelas com `n_events == 0` | 2 |
| Janelas com `n_events > 0` | 8.928 |

A taxa de falha por janela foi calculada como:

\[
fail\_rate_t = \frac{n\_failed_t}{n\_events_t}
\]

Para janelas sem eventos, foi adotado:

\[
fail\_rate_t = 0
\]

O critério oficial global foi:

\[
threshold = \mu + 2\sigma
\]

A aplicação da regra oficial resultou em:

| Parâmetro | Valor |
|---|---:|
| \(\mu\) | 0,19657958 |
| \(\sigma\) | 0,14091089 |
| `threshold` | 0,47840136 |
| Janelas críticas | 390 |
| Episódios críticos | 285 |

Com isso, o NB04 produziu a segmentação crítica oficial que será utilizada nas etapas posteriores do pipeline.

## 11. Análise de Sensibilidade com Limiar de Treino

Além da execução oficial, foi incluída uma análise de sensibilidade com limiar calculado apenas sobre os primeiros 80% da série temporal.

Essa análise foi incorporada para avaliar o impacto metodológico do uso de estatísticas globais calculadas sobre toda a série.

No cenário com limiar calculado no trecho de treino, os resultados foram:

| Parâmetro | Valor |
|---|---:|
| Fração de treino | 0,80 |
| Índice de corte do treino | 7.144 |
| \(\mu\) | 0,18158248 |
| \(\sigma\) | 0,12637807 |
| `threshold` | 0,43433863 |
| Janelas críticas | 562 |
| Episódios críticos | 388 |

Comparando com o cenário oficial global:

| Métrica | Global | Treino | Diferença |
|---|---:|---:|---:|
| `threshold` | 0,47840136 | 0,43433863 | -0,04406273 |
| Janelas críticas | 390 | 562 | +172 |
| Episódios críticos | 285 | 388 | +103 |

Esse resultado mostra que a segmentação de criticidade é sensível ao modo de estimação do limiar. Portanto, o limiar global deve ser interpretado como critério retrospectivo oficial do experimento, e não como regra diretamente equivalente a uma aplicação operacional online.

## 12. Diagnóstico do Bucket Inicial

O `bucket_id = 0` esteve presente explicitamente na série e apresentou os seguintes valores:

| Métrica | Valor |
|---|---:|
| `bucket_id` | 0 |
| `bucket_start_us` | 0 |
| `n_events` | 56.452 |
| `n_failed` | 37.082 |
| `fail_rate` | 0,65687662 |
| Crítico pelo limiar global | Sim |
| Crítico pelo limiar de treino | Sim |

O `bucket_id = 0` foi classificado como crítico tanto pelo limiar global quanto pelo limiar calculado apenas no trecho de treino.

Esse resultado confirma que o início da série não apenas existe formalmente, mas também tem impacto real sobre a estatística global e sobre a segmentação de criticidade.

## 13. Diagnóstico Complementar sem o Bucket 0

Como análise complementar, foi calculado um diagnóstico removendo apenas o `bucket_id = 0`, sem alterar a regra oficial do notebook.

Nesse cenário alternativo, os resultados foram:

| Parâmetro | Valor |
|---|---:|
| \(\mu\) | 0,19652800 |
| \(\sigma\) | 0,14083412 |
| `threshold` | 0,47819625 |
| Janelas críticas | 394 |
| Episódios críticos | 285 |

Comparando com a execução oficial:

| Cenário | Janelas críticas | Episódios críticos |
|---|---:|---:|
| Com `bucket_id = 0` | 390 | 285 |
| Sem `bucket_id = 0` | 394 | 285 |

A presença do `bucket_id = 0` aumenta ligeiramente \(\mu\) e \(\sigma\), eleva levemente o limiar e reduz o número total de janelas críticas de 394 para 390. No entanto, não altera o número total de episódios detectados no cenário global.

Esse achado indica que o bloco inicial influencia a contagem de janelas críticas, mas não produz ruptura substancial na segmentação episódica global.

## 14. Perfil dos Episódios Detectados

No cenário oficial global, os episódios detectados apresentaram as seguintes características:

| Métrica | Valor |
|---|---:|
| Episódios detectados | 285 |
| Episódios com uma janela | 227 |
| Duração média em janelas | 1,3684 |
| Duração mediana em janelas | 1,0 |
| Duração máxima em janelas | 7 |
| Duração mediana | 5 minutos |
| Duração máxima | 35 minutos |

No cenário com limiar calculado apenas no trecho de treino, os episódios apresentaram:

| Métrica | Valor |
|---|---:|
| Episódios detectados | 388 |
| Episódios com uma janela | 300 |
| Duração média em janelas | 1,4485 |
| Duração mediana em janelas | 1,0 |
| Duração máxima em janelas | 18 |
| Duração mediana | 5 minutos |
| Duração máxima | 90 minutos |

Em ambos os cenários, a maioria dos episódios é curta, com mediana de uma única janela. O limiar de treino, por ser menor, detecta mais episódios e permite episódios máximos mais longos.

## 15. Artefatos Gerados

Nesta execução, foram gerados os artefatos oficiais:

| Artefato | Finalidade |
|---|---|
| `window_5min_series_scored.parquet` | Série enriquecida com `fail_rate` e `is_critical` no cenário oficial |
| `episodes_detected.parquet` | Episódios críticos oficiais |
| `04_detect_episodes_summary.json` | Summary oficial da detecção de episódios |

Também foram gerados artefatos de sensibilidade:

| Artefato | Finalidade |
|---|---|
| `window_5min_series_scored_train_threshold.parquet` | Série classificada com limiar estimado no treino |
| `episodes_detected_train_threshold.parquet` | Episódios detectados com limiar de treino |
| `04_threshold_sensitivity_summary.json` | Summary da análise de sensibilidade |

Os artefatos oficiais continuam sendo a referência para as etapas posteriores do pipeline.

## 16. Relação com os Notebooks Posteriores

Com os episódios críticos detectados e a análise de sensibilidade documentada, o pipeline segue para:

| Notebook | Uso da saída do NB04 |
|---|---|
| NB05 | Engenharia de atributos temporais |
| NB06 | Derivação de estados e formulação supervisionada |
| NB08/NB09 | Modelagem, avaliação e visualização temporal |

A partir deste ponto, as demais etapas continuam utilizando a segmentação crítica oficial baseada no limiar global.

A análise com limiar calculado no trecho de treinamento é mantida como evidência de sensibilidade metodológica e deve ser considerada na interpretação dos resultados.

## 17. Síntese Final

O NB04 cumpre adequadamente seu papel como etapa de detecção de episódios críticos.

A execução oficial identificou 390 janelas críticas e 285 episódios críticos a partir do critério global `μ + 2σ`. A análise de sensibilidade mostrou que um limiar estimado apenas no trecho inicial da série reduz o valor do limiar e aumenta substancialmente a quantidade de janelas críticas e episódios detectados.

O diagnóstico do `bucket_id = 0` mostrou que ele é crítico e influencia a estatística global, mas sua remoção diagnóstica não altera o número total de episódios no cenário oficial.

O NB04 deve, portanto, ser considerado a etapa oficial de segmentação crítica da baseline, ficando as análises alternativas de limiar como evidência metodológica complementar.